tls/s2n_crl.h

```diff
@@ -16,16 +16,40 @@
 #pragma once
 
 #include "api/s2n.h"
+#include "utils/s2n_result.h"
 
 #include <openssl/x509v3.h>
 
+struct s2n_x509_validator;
+
 struct s2n_crl {
     X509_CRL *crl;
 };
 
+typedef enum {
+    AWAITING_RESPONSE,
+    FINISHED
+} crl_lookup_callback_status;
+
+struct s2n_crl_lookup {
+    crl_lookup_callback_status status;
+    X509 *cert;
+    uint16_t cert_idx;
+    struct s2n_crl *crl;
+};
+
+typedef int (*s2n_crl_lookup_callback) (struct s2n_crl_lookup *lookup, void *context);
+
 /* TODO: APIs are part of an unfinished CRL validation feature and are temporarily hidden
  * https://github.com/aws/s2n-tls/issues/3499 */
 struct s2n_crl *s2n_crl_new(void);
 int s2n_crl_load_pem(struct s2n_crl *crl, uint8_t *pem, size_t len);
 int s2n_crl_free(struct s2n_crl **crl);
 int s2n_crl_get_issuer_hash(struct s2n_crl *crl, uint64_t *hash);
+int s2n_crl_lookup_get_cert_issuer_hash(struct s2n_crl_lookup *lookup, uint64_t *hash);
+int s2n_crl_lookup_accept(struct s2n_crl_lookup *lookup, struct s2n_crl *crl);
+int s2n_crl_lookup_reject(struct s2n_crl_lookup *lookup);
```

Comment:
I've gone back and forth on whether "accept" and "reject" are the right words here. @maddeleine, any opinions?

Reply:
Maybe "set" and "skip"?

Reply:
An actual skip function was going to be added later, so customers could choose not to provide a CRL for a certificate and have the handshake proceed regardless.

Reply:
Maybe set, skip, and fail/reject?

Reply:
Oh I see, I wasn't thinking about "reject" actually meaning it would fail unless that CRL was required to successfully validate the chain. What's your current story for crls for certs not in the chain? I remember that was a concern in the design.

Reply:
Wait, unless I'm missing something, isn't reject only actually a failure if we end up needing the crl? https://github.com/aws/s2n-tls/blob/260e6645a0a9614c3a20a364cccc00dd1551f556/tls/s2n_crl.c#L102-L105

Reply:
Correct. Reject only fails if the CRL is needed. Skip would be to skip the revocation check for the certificate completely, even if it's in the chain of trust.

Reply:
Ah, your version of "skip" sounds like fairly scary functionality I'd want to give a scarier / more descriptive name than "skip". Like "do_not_validate" or "mark_not_revoked".

Reply:
Ah ok, that makes sense. So we could do like set, skip, and do_not_validate maybe

Reply:
Decided on set/ignore: fac5c05c2ba74efaeb0bd31856ae6f5109121201